### PR TITLE
New `LogRecord` and `Recordable` implementations.

### LINKED ISSUE #1689: [Logs API] - New LogRecord abstract class.
Specs - https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#logrecord

In order to avoid storing any data as part of API implementation, `LogRecord` can be pure virtual class with setters which would be implemented by SDK. These setters would directly serialize the data using `Recordable` implementation in format expected by exporter.

Something like -
```cpp
 class LogRecord
 {
    public:
        virtual void SetAttribute(nostd::string_view key,
                            const common::AttributeValue &value) noexcept = 0;
        virtual void SetTimeStamp(common::SystemTimestamp timestamp) noexcept = 0;
         ----
   

 };
```

### LINKED ISSUE #1691: [Logs SDK] Rename and update existing LogRecord class, modify existing Recordable interface
- To avoid conflict with the new LogRecord class, rename existing sdk::logs::LogRecord class to some better name (say sdk::logs::LogRecordData).
- Add setter for observed_timestamp in sdk:logs::Reordable and sdk::logs::LogRecordData.
- Similar to Span's Recordable, add a getter method for LogRecordData in Recordable class. This can be eventually implemented by all it's derived implementation (not in scope for now).
```cpp
  /**
   * Get the LogRecordData object for this Recordable.
   *
   * @return LogRecordData*
   */

  virtual explicit operator LogRecordData *() const { return nullptr; }

```

This would let us avoid having `ReadWriteLogRecord` class. This method can be potentially used inside `LogRecordProcessor` to filter the logs based on it's contents (not in scope for now).

### PR DESCRIPTION
**Pull request by owent**
Signed-off-by: WenTao Ou <admin@owent.net>

Fixes #1691 
Fixes #1689 

## Changes

**There are a lot break changes in this PR.But it's easy to migrate if users do not implement their own Logger and Exporter.**

+ Removed deprecated `Logger::Log` APIs.
  + All APIs with the old `name` field are removed now.The `name` field is ognored from v1.4.0.
+ New `LogRecord`, `ReadableLogRecord`, `ReadWriteLogRecord`
  + https://github.com/open-telemetry/opentelemetry-specification/blob/v1.15.0/specification/logs/api.md#logrecord
  + https://github.com/open-telemetry/opentelemetry-specification/blob/v1.15.0/specification/logs/sdk.md#additional-logrecord-interfaces
+ New `logs::Recordable` to collect data from the all in one `LogRecord`.
  + #1691 
+ Add a simple `EmitEvent` function which will be use to implement #1687 and #1688 in the future.
  + https://github.com/open-telemetry/opentelemetry-specification/blob/v1.15.0/specification/logs/api.md#emit-event
  + It's not finished yet, and will be continued in another PR later.
+ Replace `std::unique_ptr<T>` with `nostd::unique_ptr<T>` in logs.

For significant contributions please make sure you have completed the following items:

* [x] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

File: sdk/include/opentelemetry/sdk/logs/read_write_log_record.h
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/logs/read_write_log_record.h:168]: So, these LogRecord fields (severity, body, attributes) would be copied/stored in the SDK before exporting them, unlike earlier when they were directly serialized in the exporter's format ?

Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h:168]: Yes, according to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#logrecord , [LogRecords](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/api.md#emit-logrecord) MUST be able to set these following fields.
So I think we can only store them into ReadWriteLogRecord, and we can just store `string_view` and integers here to reduce copy.

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h:168]: Theonly advantage I see with having ReadableLogRecord, and ReadWriteLogRecord is that we can now easily define the getter methods for logRecord as required by specs. But I am not seeing the use-case of these getter methods. For traces too, we deviated from specs by not have read/write spans and directly uses recordable to serialize the data as a tradeoff for performance.

Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h:168]: `ReadableLogRecord` and `ReadWriteLogRecord` is defined in SDK, while `LogRecord` is defined in API. They are just like `Span` in API and SDK, and `logs::Recordable` is just like the `trace::Recordable`.
I believe which cost more is store data into `proto::v1::LogRecord` in OTLP and convert data into json in ETW and ES exporters.To reduce the cost, I try to only keep string_view and AttributeValue in LogRecord classes, these are all just a reference to the real data without lifetime(There is a except, attributes are store in a `unordered_map<std::string, AttributeValue>` now, maybe we can use a reference to `KeyValueIterable` instead of using `unordered_map`, but it may make it difficult to set attribute one by one and add `event.name` when emit a record with event), The getter methods are used to pick datas from LogRecord and set them into `Recordable` now.

Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h:168]: I think it easier to add or remove fields if log content changes in the future with `LogRecord` in API.

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h]: Storing non-owning objects can have undesirable issues if these objects get out of scope before they get serialized by the exporter. I am still not seeing the benefit of storing them for implementing getter methods unless there is a use case where these getters would be used as public API.

Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h]: The exporter should create it's own `Recordable` object from `ReadableLogRecord` or `ReadWriteLogRecord` and the `Recordable` of exporter should own objects, just like the old implementation.The mainly different between old and new implementation is the old implementation set data fields one by one and the new implementation collect all data into `ReadWriteLogRecord` the lifetime and data copy times will not change, just like we talked before in #1691 .
`ReadableLogRecord` is not necessary but I think it's better to follow [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.15.0/specification/logs/sdk.md#readablelogrecord). Do you mean we can delete `ReadWriteLogRecord` and just keep `ReadWriteLogRecord` ?

**Comment by lalitb** [comment]
> All APIs with the old name field are removed now.

Do you think we can keep the existing API intact (maybe renaming to `Logger::EmitLog(..., ..., ...)`)  as mentioned in #1688. As per the logging API specs, the newly introduced interface to emit the log-records is only for Appenders. The specs doesn't enforce it for instrumentation library which makes sense because most of the languages (except C++) have logging api as part of language specs. The existing API can continue using direct `Recordable` interface to send the logs to exporter, without creating LogRecord object. Sometime, instrumentation library author would prefer to write a single line of code to emit logs.

And then the new interface can build the `Recordable` through `ReadWriteLogRecord` as you are doing right now in this PR.

**Comment by owent** [comment]
> intact



> > All APIs with the old name field are removed now.
> 
> Do you think we can keep the existing API intact (maybe renaming to `Logger::EmitLog(..., ..., ...)`) as mentioned in #1688. As per the logging API specs, the newly introduced interface to emit the log-records is only for Appenders. The specs doesn't enforce it for instrumentation library which makes sense because most of the languages (except C++) have logging api as part of language specs. The existing API can continue using direct `Recordable` interface to send the logs to exporter, without creating LogRecord object. Sometime, instrumentation library author would prefer to write a single line of code to emit logs.
> 
> And then the new interface can build the `Recordable` through `ReadWriteLogRecord` as you are doing right now in this PR.

Sorry, I'm poor in English and perhaps I didn't make it clear. The `name` field means All `Logger::Log` with `name` parameter, this argument is ignored, for example `OPENTELEMETRY_DEPRECATED_MESSAGE("name will be removed in the future") void Log(Severity severity, nostd::string_view name, nostd::string_view message) noexcept`. We have discussed at #1383 before, and I think it's long enough to leave time for users to migrate now.

For users, we can still use the old `Logger::Log` APIs to emit logs, if we allow processor or exporters to use `Recordable` interface to send the logs without `LogRecord`, we can make `Recordable` derive from `LogRecord`. But if we changes or add more fields into `LogRecord` , we have to modify all `Recordable` implementations. Do you think it will be more difficult to maintain and upgrade for users?

File: sdk/include/opentelemetry/sdk/logs/read_write_log_record.h
Comment (lalitb) [reply sdk/include/opentelemetry/sdk/logs/read_write_log_record.h]: > does it means these getters would be able to be used as public API?

Yes, these getters should be available as public API.

**Comment by lalitb** [comment]
> > intact
> 
> > > All APIs with the old name field are removed now.
> > 
> > 
> > Do you think we can keep the existing API intact (maybe renaming to `Logger::EmitLog(..., ..., ...)`) as mentioned in #1688. As per the logging API specs, the newly introduced interface to emit the log-records is only for Appenders. The specs doesn't enforce it for instrumentation library which makes sense because most of the languages (except C++) have logging api as part of language specs. The existing API can continue using direct `Recordable` interface to send the logs to exporter, without creating LogRecord object. Sometime, instrumentation library author would prefer to write a single line of code to emit logs.
> > And then the new interface can build the `Recordable` through `ReadWriteLogRecord` as you are doing right now in this PR.
> 
> Sorry, I'm poor in English and perhaps I didn't make it clear. The `name` field means All `Logger::Log` with `name` parameter, this argument is ignored, for example `OPENTELEMETRY_DEPRECATED_MESSAGE("name will be removed in the future") void Log(Severity severity, nostd::string_view name, nostd::string_view message) noexcept`. We have discussed at #1383 before, and I think it's long enough to leave time for users to migrate now.
> 
> For users, we can still use the old `Logger::Log` APIs to emit logs, if we allow processor or exporters to use `Recordable` interface to send the logs without `LogRecord`, we can make `Recordable` derive from `LogRecord`. But if we changes or add more fields into `LogRecord` , we have to modify all `Recordable` implementations. Do you think it will be more difficult to maintain and upgrade for users?

Thanks for explanation. So to summarise 
 - We can remove all the deprecated `Logger::Log()` methods ( i.e. those containing `name` ), as we have given enough time for users to migrate.
 -  Rest of the existing methods can be (probably ?) renamed to `Logger::EmitLog()`, and they will keep using `Recordable` interface (which is derived from `LogRecord` class).

This looks fine to me. Log Data model is already stable, I don't think LogRecord would change any further, definitely not after log API is also made stable. So, there shouldn't be upgrade issues with users.

**Comment by owent** [comment]
> > > intact
> > 
> > 
> > > > All APIs with the old name field are removed now.
> > > 
> > > 
> > > Do you think we can keep the existing API intact (maybe renaming to `Logger::EmitLog(..., ..., ...)`) as mentioned in #1688. As per the logging API specs, the newly introduced interface to emit the log-records is only for Appenders. The specs doesn't enforce it for instrumentation library which makes sense because most of the languages (except C++) have logging api as part of language specs. The existing API can continue using direct `Recordable` interface to send the logs to exporter, without creating LogRecord object. Sometime, instrumentation library author would prefer to write a single line of code to emit logs.
> > > And then the new interface can build the `Recordable` through `ReadWriteLogRecord` as you are doing right now in this PR.
> > 
> > 
> > Sorry, I'm poor in English and perhaps I didn't make it clear. The `name` field means All `Logger::Log` with `name` parameter, this argument is ignored, for example `OPENTELEMETRY_DEPRECATED_MESSAGE("name will be removed in the future") void Log(Severity severity, nostd::string_view name, nostd::string_view message) noexcept`. We have discussed at #1383 before, and I think it's long enough to leave time for users to migrate now.
> > For users, we can still use the old `Logger::Log` APIs to emit logs, if we allow processor or exporters to use `Recordable` interface to send the logs without `LogRecord`, we can make `Recordable` derive from `LogRecord`. But if we changes or add more fields into `LogRecord` , we have to modify all `Recordable` implementations. Do you think it will be more difficult to maintain and upgrade for users?
> 
> Thanks for explanation. So to summarise
> 
> * We can remove all the deprecated `Logger::Log()` methods ( i.e. those containing `name` ), as we have given enough time for users to migrate.
> * Rest of the existing methods can be (probably ?) renamed to `Logger::EmitLog()`, and they will keep using `Recordable` interface (which is derived from `LogRecord` class).
> 
> This looks fine to me. Log Data model is already stable, I don't think LogRecord would change any further, definitely not after log API is also made stable. So, there shouldn't be upgrade issues with users.

Thanks, I will push another commit later to change the relationship between `Recordable` and `LogRecord`.

I also have a idea about the new `Logger::EmitLog()` APIs, how about using variable template when implement `Logger::EmitLog()`, and call specify `LogRecord::SetXXX` by type of argument?When we get `trace::TraceId` , call `LogRecord::SetTraceId()`,call `LogRecord::SetAttribute()` when we get any type that can be converted to `common::KeyValueIterable` and so on. So that we don't need to implement so many overload `Logger::EmitLog()` functions.

**Comment by lalitb** [comment]
> I also have a idea about the new Logger::EmitLog() APIs, how about using variable template when implement Logger::EmitLog(), and call specify LogRecord::SetXXX by type of argument?When we get trace::TraceId , call LogRecord::SetTraceId(),call LogRecord::SetAttribute() when we get any type that can be converted to common::KeyValueIterable and so on. So that we don't need to implement so many overload Logger::EmitLog() functions.

**Comment by owent** [comment]
> > I also have a idea about the new Logger::EmitLog() APIs, how about using variable template when implement Logger::EmitLog(), and call specify LogRecord::SetXXX by type of argument?When we get trace::TraceId , call LogRecord::SetTraceId(),call LogRecord::SetAttribute() when we get any type that can be converted to common::KeyValueIterable and so on. So that we don't need to implement so many overload Logger::EmitLog() functions.
> 
> Aren't variable templates part of C++14 ?

No, it's C++11. See https://en.cppreference.com/w/cpp/compiler_support/11

**Comment by lalitb** [comment]
> > > I also have a idea about the new Logger::EmitLog() APIs, how about using variable template when implement Logger::EmitLog(), and call specify LogRecord::SetXXX by type of argument?When we get trace::TraceId , call LogRecord::SetTraceId(),call LogRecord::SetAttribute() when we get any type that can be converted to common::KeyValueIterable and so on. So that we don't need to implement so many overload Logger::EmitLog() functions.
> > 
> > 
> > Aren't variable templates part of C++14 ?
> 
> No, it's C++11. See https://en.cppreference.com/w/cpp/compiler_support/11

Oh variadic templates :) Not sure if that would look good at public API surface. I would prefer overloaded methods, but would leave it to you.

**Comment by owent** [comment]
@lalitb Could you please take some time to review this PR again when you have time?The `Recordable` is derived from `LogRecord` now.
I think I can rename the other `Logger::Log` into `Logger::EmitLog` in another PR and maybe we can also keep the old `Log` APIs for serveral versions to let users to migrate.

File: api/include/opentelemetry/logs/logger.h
Comment (lalitb) [inline api/include/opentelemetry/logs/logger.h:56]: Not for immediate attention, but just to update that Log/Event API specs has changed again :( - https://github.com/open-telemetry/opentelemetry-specification/pull/2941/

We can accommodate these changes separately in another PR. Will create a task for that.

File: api/include/opentelemetry/logs/logger.h
Comment (lalitb) [inline api/include/opentelemetry/logs/logger.h:135]: Please confirm if the existing methods wouldn't use LogRecord, but directly use Recordable interface ?

File: api/include/opentelemetry/logs/logger.h
Comment (owent) [reply api/include/opentelemetry/logs/logger.h]: OK, I removed `EmitEvent` and will add it back into `EventLogger` later when implementing #1687 . We can just modify #1687 to follow the changes.

File: api/include/opentelemetry/logs/logger.h
Comment (owent) [reply api/include/opentelemetry/logs/logger.h:97]: `Recordable` derive from `LogRecord`. It's a break change in API here, the existing old `UserLogger::Log` must be replaced by `UserLogger::CreateLogRecord` and `UserLogger::EmitLogRecord`, or it won't compile.(Just like the changes in `etw::Logger`).
And the existing exporters or processes use the old `Recordable` still work after add a new override `Recordable::SetObservedTimestamp` function.

Comment (lalitb) [reply api/include/opentelemetry/logs/logger.h:97]: Sorry for the delay. Let me go through all the changes in next couple of days, and would complete the review by then.

Comment (owent) [reply api/include/opentelemetry/logs/logger.h:97]: Thanks, please let me know if there is any misunderstanding about specification.

File: sdk/include/opentelemetry/sdk/logs/batch_log_record_processor.h
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/logs/batch_log_record_processor.h:44]: Why do we need to switch to `nostd::unique_ptr` ?

File: sdk/include/opentelemetry/sdk/logs/logger.h
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/logs/logger.h:34]: Why do we need to switch to `nostd::unique_ptr` here, and in general in most of other places in SDK ?

File: api/include/opentelemetry/logs/logger.h
Comment (lalitb) [reply api/include/opentelemetry/logs/logger.h:97]: No, I think the changes are much aligned to the specs, and look better. I would have some minor comments only.

File: sdk/include/opentelemetry/sdk/logs/logger.h
Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/logger.h:34]: Isn't it has better ABI compatibility? I think we should better use `nostd::unique_ptr` everywhere.

Comment (lalitb) [reply sdk/include/opentelemetry/sdk/logs/logger.h:34]: We don't claim ABI compatibility at SDK level. And so have been using `std::` classes throughout SDK ( in both traces and metrics). Only API interfaces implemented by SDK use `nostd::` classes.

Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/logger.h]: Thanks for the explaination, I restore `nostd::unique_ptr` to `std::unique_ptr` in sdk and exporters now.

File: sdk/src/logs/logger.cc
Comment (lalitb) [inline sdk/src/logs/logger.cc:57]: Oh I didn't realize this would need conversion from nostd::unique_ptr to std::unique_ptr.

**Review by lalitb — COMMENTED** [review]
Have some minor comments. In general looks in good shape. Need to review otlp exporter part to finish the review, will be doing that tomorrow.

File: api/include/opentelemetry/logs/logger.h
Comment (owent) [reply api/include/opentelemetry/logs/logger.h:95]: Should `observerd_timestamp` always be set to the time when log record is created?

File: api/include/opentelemetry/logs/logger.h
Comment (owent) [reply api/include/opentelemetry/logs/logger.h:141]: According to specification, log should be emitted by `Logger::EmitLogRecord`, not `Logger::Log`.

File: sdk/src/logs/logger.cc
Comment (owent) [reply sdk/src/logs/logger.cc:57]: Because API create `nostd::unique_ptr` but sdk and exporters use `std::unique_ptr`, as we talked before at https://github.com/open-telemetry/opentelemetry-cpp/pull/1766#discussion_r1037982249_

File: sdk/include/opentelemetry/sdk/logs/exporter.h
Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/exporter.h]: Thanks and removed.

File: sdk/include/opentelemetry/sdk/logs/exporter.h
Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/exporter.h]: Thanks and sorry I forget to update these comments.

File: sdk/include/opentelemetry/sdk/logs/processor.h
Comment (owent) [reply sdk/include/opentelemetry/sdk/logs/processor.h]: Thanks and sorry I forget to update these comments.

File: sdk/src/logs/logger.cc
Comment (owent) [reply sdk/src/logs/logger.cc]: Thanks and removed.

File: api/include/opentelemetry/logs/logger.h
Comment (lalitb) [reply api/include/opentelemetry/logs/logger.h:95]: Yes, I think we can set it internally as current timestamp instead of making it part of Logger::Log() API.

File: exporters/etw/include/opentelemetry/exporters/etw/etw_logger.h
Comment (lalitb) [inline exporters/etw/include/opentelemetry/exporters/etw/etw_logger.h:190]: This will create a intermediate storage for LogRecord object even when we invoke Logger:Log(). We can keep it as of now, I will try to see it later if we can optimize it.

File: sdk/src/logs/readable_log_record.cc
Comment (lalitb) [inline sdk/src/logs/readable_log_record.cc:31]: nit - is this comment valid, as we seems to be using shared static instrumentation scope below.

File: sdk/src/logs/read_write_log_record.cc
Comment (lalitb) [inline sdk/src/logs/read_write_log_record.cc:145]: nit - use OPENTELEMETRY_LIKELY_IF ?

**Review by lalitb — APPROVED** [review]
LGTM. with few minor comments. 
We can wait for couple of days if there are more reviews before merging.

Thanks for the work, it is nicely done :)

File: sdk/src/logs/readable_log_record.cc
Comment (owent) [reply sdk/src/logs/readable_log_record.cc:31]: The shared default instrumentation scope I mean is a shared instrumentation scope between all of the components of SDK, includes trace and metrics, not just for logs.

File: sdk/src/logs/read_write_log_record.cc
Comment (owent) [reply sdk/src/logs/read_write_log_record.cc:145]: Yes and thanks, I will change it tomorrow.

**Pull request merged**
